### PR TITLE
ExpectThunk proposal

### PR DESCRIPTION
**Pull request by obj-p**
Hey all, this is just a thought that I've been exploring previously with ActionCreators and just open for discussion if it has a place somewhere in ReSwift. Here's and example of the previous rendition:
https://github.com/jjgp/ReddiX-iOS/blob/master/UNITests/Actions/ListingsActionsTests.swift

To make Thunks more testable especially when facing async things like mocking requests with MockingJay or OHHTTPStubs I created an `ExpectThunk` that is a subclass of `XCTestExpectation`.

If this is to be then there would have to be a subspec or other podspec for consumers to be able to pod install into their tests. 

Cheer,

### CODE REVIEW

File: ReSwift-ThunkTests/Tests.swift [hunk 1]
```diff
@@ -58,4 +59,36 @@ class Tests: XCTestCase {
...
+            XCTAssertNotNil(getState())
+            DispatchQueue.global().asyncAfter(deadline: DispatchTime.now() + 0.5) {
+                dispatch(AnotherFakeAction())
+                XCTAssertNotNil(getState())
+                XCTAssertNil(getState())
+            }
+            dispatch(FakeAction())
+        }
+        
+        let expect = ExpectThunk<FakeState>(thunk)
```
Comment (obj-p) [inline ReSwift-ThunkTests/Tests.swift]: Here's the result of the proposal! Basically a chain-able expectation much like redux saga test plan

**Comment by mjarvis** [comment]
Hey, this look great! I can see adding this as a sub-spec to making testing thunks easier.
One improvement I can see making is adding another `dispatches` function that takes a closure `(Action) -> Void` to be able to manually match and assert on actions outside of the `Equatable` restriction (maybe some action is difficult to make equatable).

**Comment by obj-p** [comment]
@[mjarvis](/mjarvis) great feedback! I will incorporate and leave this open a tad longer in case there is any other feedback. Then I will close and make some refactors and reopen the PR.

**Comment by mjarvis** [comment]
@jjgp you are welcome to leave the PR open and make changes in-place. We'll use a squash & merge to merge when its ready so any WIP commits will get squashed out.

File: ReSwiftExpectThunk.podspec [hunk 2]
```diff
@@ -0,0 +1,29 @@
+Pod::Spec.new do |spec|
```
Comment (mjarvis) [inline ReSwiftExpectThunk.podspec]: I think this would probably be better suited as a subspec rather than a separate pod inside the same repo.
See https://guides.cocoapods.org/syntax/podspec.html#subspec and also https://guides.cocoapods.org/syntax/podspec.html#default_subspecs for omitting this from the default.

I would create two subspecs `Core`, and `Expect`, then set `Core` as the `default_subspec`

Comment (obj-p) [reply ReSwiftExpectThunk.podspec]: I'm all for making it a subspec. I went with this route first to test it in a demo app. I will attempt a subspec next. Will report back on this.

Second thing I would be for adding a blurb in the README.md if that's cool.

Comment (mjarvis) [reply ReSwiftExpectThunk.podspec]: 👍 👍

Comment (obj-p) [reply ReSwiftExpectThunk.podspec]: I believe I got the subspec down. I have a test project here: https://github.com/jjgp/ExpectThunkDemo

I'd appreciate a sanity check since I did run into CocoaPods fun with a sigabrt at test time, however, I managed to resolve it.

Will work on the blurb in a bit.

**Comment by obj-p** [comment]
~travis is failing and I'll take a look in a bit~

File: ReSwift-ThunkTests/ExpectThunk.swift [hunk 3]
```diff
@@ -8,7 +8,6 @@
 
 import XCTest
 import ReSwift
-import ReSwiftThunk
```
Comment (obj-p) [inline ReSwift-ThunkTests/ExpectThunk.swift]: ahhh, it's a warning to leave this in when integrating into a 3rd party. This is why Travis is failing. I will think about it what to do here

File: ReSwiftExpectThunk.podspec (hunk 2 above)
Comment (obj-p) [reply ReSwiftExpectThunk.podspec]: sigabrt will show up if the test target is a child target in the Podfile:

```
target 'ExpectThunkDemo' do
  # Comment the next line if you're not using Swift and don't want to use dynamic frameworks
  use_frameworks!

  # Pods for ExpectThunkDemo
  pod 'ReSwiftThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  target 'ExpectThunkDemoTests' do
    inherit! :search_paths
    pod 'ReSwiftThunk/ExpectThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  end
end
```

**Comment by xavierLowmiller** [comment]
Are there any news on this?

I don't mean to be pushy, but we'd love to use this.

Or, since this is 'just' for tests, do you think it's viable to target this branch in our project?

**Comment by obj-p** [comment]
Appreciate the interest! I’ll get back to it today. It should be near complete. After I double check I’ll ping for some reviews. Cheers,

**Comment by obj-p** [comment]
Hey @mjarvis and @DivineDominion! Mind reviewing this PR when the chance presents? A sample project testing the subspec is [ExpectThunkDemo](https://github.com/jjgp/ExpectThunkDemo). 

The one item that I have slight hesitation on is the subspec and specification in the Podfile. It seems the only way to use a subspec and the core spec is to ensure that the Target and its Target-Tests are not nested. For example, this works:

```
target 'ExpectThunkDemo' do
  use_frameworks!

  # Pods for ExpectThunkDemo
  pod 'ReSwiftThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
end

Target 'ExpectThunkDemoTests' do
  pod 'ReSwiftThunk/ExpectThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
end
``` 

While this specification has issues (sigabrt due to missing symbol): 

```
target 'ExpectThunkDemo' do
  use_frameworks!

  pod 'ReSwiftThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  target 'ExpectThunkDemoTests' do
    inherit! :search_paths
    pod 'ReSwiftThunk/ExpectThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  end
end
```

Appreciate it!

File: ReSwiftThunk.podspec [hunk 4]
```diff
@@ -20,7 +20,19 @@ Pod::Spec.new do |spec|
...
     :git => "https://github.com/ReSwift/ReSwift-Thunk.git",
     :tag => spec.version.to_s }
-  spec.source_files = "ReSwift-Thunk"
+
+  spec.subspec "Core" do |sp|
+    sp.source_files = "ReSwift-Thunk"
+  end
+
+  spec.subspec "ExpectThunk" do |sp|
+    sp.dependency "ReSwiftThunk/Core"
```
Comment (mjarvis) [inline ReSwiftThunk.podspec:29]: If you remove this dependency, perhaps the inherited podfile example will work? I think it might be double-importing `Core` when `inherit` is included in the test target pod specifications.

Edit: Though, really, cocoapods dependency stuff should really handle this automatically...

Comment (obj-p) [reply ReSwiftThunk.podspec:29]: Good thought, however, I wasn't able to get it to build this way. I'm attempting with CocoaPods 1.6.1. It may be an open issue with CocoaPods [7195](https://github.com/CocoaPods/CocoaPods/issues/7195) having child targets with subspecs. I'll follow up there, it does seem splitting the targets is the workaround in the Podfile at the moment.

**Comment by DivineDominion** [comment]
I'm by no means a CocoaPods expert and almost never use it. Is it common to nest test targets inside the library/app target, or just an edge case?

Totally missed this whole PR -- looks very cool!

**Comment by obj-p** [comment]
So CocoaPods defaults to nested targets on a `pod init`. That being said having them separate is common and often used as a workaround for certain issues. Even documentation such as RxSwift shows them as separate:

```
# Podfile
use_frameworks!

target 'YOUR_TARGET_NAME' do
    pod 'RxSwift',    '~> 4.0'
    pod 'RxCocoa',    '~> 4.0'
end

# RxTest and RxBlocking make the most sense in the context of unit/integration tests
target 'YOUR_TESTING_TARGET' do
    pod 'RxBlocking', '~> 4.0'
    pod 'RxTest',     '~> 4.0'
end
```

**Comment by DivineDominion** [comment]
@jjgp Do you need any assistance to make things work? I'd very much like to integrate this rather sooner than later :)

Also, what do you think about exposing a record of dispatched actions or similar, so one can write e.g. 3 `dispatches` assertions _and_ make sure only `3` actions have been dispatched in total (e.g. there's nothing on top of that).

**Comment by obj-p** [comment]
Appreciate it @DivineDominion! I'm going to get a version of that `dispatches` in over the weekend. I had a few other thoughts as well.

The only thing I'm unsure about is CocoaPods. I tend to lean that breaking it into separate targets is acceptable for the end user. The issue sited on CocoaPods repo has been moved to a further milestone and perhaps I will have to contribute there.

I'll report back by the end of the weekend.

**Comment by obj-p** [comment]
@DivineDominion currently I do not enforce the order of getStates within the dispatches/getStates chain. I haven't really thought about it, do you have an opinion of it?

**Comment by obj-p** [comment]
Screen Shot 2019-05-18 at 2 37 33 PM

Work in progress, I have included the `dispatches` record as well as built in a way for `ExpectThunk` to mark parts of the chain that were not reached.

**Comment by obj-p** [comment]
Looks like the following Podfile works as well:
```
platform :ios, '12.1'

target 'ExpectThunkDemo' do
  pod 'ReSwiftThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  target 'ExpectThunkDemoTests' do
    pod 'ReSwiftThunk/ExpectThunk', :git => 'git@github.com:jjgp/ReSwift-Thunk.git', :branch => 'expect-thunk-proposal'
  end
end
```

Moral of the story is that if your using the `use_frameworks!` option then it will SIGABRT in the tests probably because the framework isn't being dynamically linked or something like that.

Here is a branch where it works with the above Podfile sans `use_frameworks!`:
https://github.com/jjgp/ExpectThunkDemo/compare/separate-targets

Looking back at this comment, that is the distinction too: https://github.com/ReSwift/ReSwift-Thunk/pull/19#issuecomment-484124755

I'm leaning more towards this is fine CocoaPods wise as the user has a few options and the framework linking seems to make sense.

**Comment by obj-p** [comment]
updated failure descriptions:
Screen Shot 2019-05-18 at 7 40 34 PM

**Comment by DivineDominion** [comment]
> Moral of the story is that if your using the `use_frameworks!` option then it will SIGABRT in the tests probably because the framework isn't being dynamically linked or something like that.

Do I read this correctly: this makes it impossible to use `use_Frameworks!` for devs who use this sub-spec of the pod?

**Comment by obj-p** [comment]
`use_frameworks!`may still be used as long as the testing target is defined outside of the source target in the Podfile like in this demo project 
https://github.com/jjgp/ExpectThunkDemo/blob/master/Podfile

It seems if the test target is nested in the source target and it declares `use_frameworks!` then the issue presents itself.

It could be related to this issue with CocoaPods as listed here: https://github.com/CocoaPods/CocoaPods/issues/7195
~~Although I don’t see others using `use_frameworks` there.~~ Actually: https://github.com/CocoaPods/CocoaPods/issues/7195#issuecomment-341949444

Another way I’ve seen test helpers implemented is as a separate CocoaPod entirely such as RxTest. I would have to test, however, I’m inclined to think it wouldn’t have the issue seen here.

**Comment by DivineDominion** [comment]
Cool! I've tested the code in my app and really like the test helper. So if someone with more CocoaPods experience (maybe @mjarvis or @danielmartinprieto) accept the CocoaPods side of things, I think we should merge this 👍

**Comment by DivineDominion** [comment]
@mjarvis @danielmartinprieto Does any one of you have an opinion regarding the pod subspec mentioned [above](https://github.com/ReSwift/ReSwift-Thunk/pull/19#issuecomment-493701660)?

**Comment by mjarvis** [comment]
I'm okay with a bit of awkwardness as long as:
1. It doesn't affect installation of the main project
2. Its well documented

**Comment by obj-p** [comment]
I will update the README to make it more clear. Also looking to the future I’ll look into Swift Package Manager and it may make for a cleaner integration!

**Comment by obj-p** [comment]
Just updated the README to make it explicit to not nest the targets if using the subspec 😄

**Comment by DivineDominion** [comment]
Great 👍

Could you add your addition to the top of the changelog? (Sorry I'm serving all this info piecemeal and asking you to always edit one more thing :))

**Comment by obj-p** [comment]
Just added a blurb to the changlog. I hope I put it in the expected place 😄

**Comment by DivineDominion** [comment]
Just noticed you enabled editing by maintainers, that's good :) I edited in some details. Thanks for all this!

**Review by DivineDominion — APPROVED** [review]
